### ISSUE TITLE
[RFC] Should DDP support custom reduction logic for registered module buffers?

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#63041]
## 🚀 Feature
Support custom logic (i.e. collective reduction) for registered module buffers as part of DDP training. Currently, when buffers are registered as part of a model, DDP will broadcast them from rank 0 or do nothing if `self.broadcast_buffers=False`. However, we have seen some use cases (such as [here](https://discuss.pytorch.org/t/how-to-prevent-some-buffers-from-being-broadcasted-while-having-broadcast-buffers-true/128957)), as well as a request internally.

## Motivation

The main motivation is that users would like to register certain tensors as buffers so that they are saved as part of `state_dict` to be a part of model checkpoint, but semantically `broadcast` doesn't make sense for these buffers. One example is tracking the no. of positive/negative examples a rank has seen and using that in computing a weighted loss, where we may want to `allreduce` this ratio instead of broadcasting rank 0's.

## Pitch

A communication hook (similar to existing gradient communication hooks) that allow users to specify custom buffer reductions. It will default to `broadcast` to maintain compatibility with the current version. Because different buffers may require different reduction, it may have to be a per-buffer lambda specified by the user:

```
class MyModel():
    def __init__(self):
        self.my_buf = register_buffer(...)

ddp = DDP(MyModel())
ddp.register_buffer_comm_hook(ddp.my_buf, lambda buf: dist.all_reduce(buf))
```

## Performance consideration
We currently coalesce all buffers and broadcast them instead of broadcasting them individually, which reduces the no. of collective comms. we launch and allows us to do it all in one go. With these sort of custom hooks, it is unlikely we will be able to do such a coalesced operation because each buffer could have different reduction logic, which might impact the performance if user chooses to use this feature.

## Alternatives

Instead of hooks which gives more control, it might be simpler to offer setting a strategy on how to reduce each buffer, such as `broadcast`, `allreduce`, etc. This may be simpler but is less configurable.

**Comment by pritamdamania87** [comment]
> We currently coalesce all buffers and broadcast them instead of broadcasting them individually, which reduces the no. of collective comms. we launch and allows us to do it all in one go. With these sort of custom hooks, it is unlikely we will be able to do such a coalesced operation because each buffer could have different reduction logic, which might impact the performance if user chooses to use this feature.

Can we instead pass all the buffers to the hook in a parameter of type Dict[str, Tensor] where the key is the buffer name and the Tensor is the actual buffer? In that case, we can do more efficient coalesced broadcast in the default implementation of the hook. Even the user can use coalesced options for handling different types of buffers.

**Comment by rohan-varma** [comment]
> Can we instead pass all the buffers to the hook in a parameter of type Dict[str, Tensor] where the key is the buffer name and the Tensor is the actual buffer? In that case, we can do more efficient coalesced broadcast in the default implementation of the hook. Even the user can use coalesced options for handling different types of buffers.

Makes sense, I think `Dict[str, Tensor]` is pretty much what we get from `named_buffers` so we can pretty much just pass that into the lambda. 

The question would be what to do with buffers that the user doesn't have custom reduction for, i.e. 
```
def custom_buffers_hook(named_buffers):
    for name, buf in named_buffers:
        if name == "foo": custom_func(buf)
    # user wants to use default broadcast for reset of buffers
```

It seems like the only option is that the user must add their own `broadcast`/`broadcast_coalesced` logic to the rest of their `named_buffers`, i.e. we can't really do it ourselves since we don't know which buffers they are using in their custom hook anymore.

**Comment by pritamdamania87** [comment]
> The question would be what to do with buffers that the user doesn't have custom reduction for, i.e.

I think my proposal was to have the user control reduction for all buffers or none. So when the user registers a `custom_buffers_hook` they are responsible for reducing all the buffers. This is similar to the ddp comm hook where the user controls all grad reductions.

In the use case we are targeting, if the user only cares about a few parameters. We could have a helper function which takes rest of the named_buffers as `Dict[str, Tensor]` and does the default broadcast_coalesced. Then users in their custom hook handle their specific parameters, remove them from the named_buffers list and pass the rest to the function we provided.

**Comment by zhaojuanmao** [comment]
some use cases only need to sync buffers once per epoch, providing helper functions/APIs possibly is a flexible way to help users sync buffers https://github.com/pytorch/pytorch/issues/30718

**Comment by rohan-varma** [comment]
@zhaojuanmao Thanks for raising that, it seems like it can be done with the hook proposed here. It will mostly be the users responsibility - they can pass in a `state` object that is modified appropriately, and then used to check if the reduction needs to run.
